### PR TITLE
Update JwtBearer, WsFed, and OIDC handlers to use identity model 7

### LINKED ISSUE #49469: Move aspnetcore to leverage JsonWebToken and JsonWebTokenHandler
## Background and Motivation

Improvements in `JsonWebToken` and `JsonWebTokenHandler` have been made in `Microsoft.IdentityModel`, which include a 30% performance improvement over `JwtSecurityToken` which is currently used in ASP.NET today. In later versions of Microsoft.IdentityModel 7.x (before .NET8 RC1), we will enable AOT support by having fully trimmable assemblies in `Microsoft.IdentityModel` and remove the dependency of Newtonsoft, enabling a smaller dll for AOT.

`Microsoft.IdentityModel` offers two generations of JSON web token (JWT) handling, which are in two assemblies:

- `System.IdentityModel.Tokens.Jwt` is the old generation. Notable types are `JwtSecurityToken` and `JwtSecurityTokenHandler`. This is the assembly currently used by ASP.NET Core.
- `Microsoft.IdentityModel.Tokens.JsonWebToken` is the next generation. It offers `JsonWebToken` and `JsonWebTokenHandler` with: 

    - Improved performance (30%)

    - Better resilience: IdentityModel will fetch and maintain the OIDC metadata and uses its last known good state (repair item from March 2020 outage)

    - Defense in depth: IdentityModel provides additional [AAD key issuer validation](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2134) protection

    - Support for async token validation, returning a `TokenValidationResult` rather than throwing

`Microsoft.IdentityModel` also has two abstractions for Token Handlers:
-	`ISecurityTokenValidator` is the old generation.
-	`TokenHandler` (abstract class) is the next generation and offers asynchronous token validation.

The following assemblies have a dependency on `JwtSecurityToken`, or `JwtSecurityTokenHandler`:
-	`Microsoft.AspNetCore.Authentication.JwtBearer`,
-	`Microsoft.AspNetCore.Authentication.OpenIdConnect`,
-	`Microsoft.AspNetCore.Authentication.WsFederation`, 
-	`Microsoft.AspNetCore.Authentication`

## Proposed API

We introduce a new boolean, `UseTokenHandlers`, in the JwtBearer and WsFederation options to enable developers to decide whether the access token validation will be done with the new `TokenHandlers` (more performant, resilient and async) or with the legacy `SecurityTokenValidators`. We also expose the list of `TokenHandlers` used to validate the token. Developers can decide to add their own `TokenHandlers` (token type) for each protocol (JwtBearer/WsFed). By default `JwtBearerOptions.TokenHandlers` contains an instance of the `JsonWebTokenHandler` and `WsFederationOptions.TokenHandlers` contains handlers for SAML1, SAML2, and `JsonWebTokenHandler`.

[PR for reference](https://github.com/dotnet/aspnetcore/pull/48966)

Additions in `JwtBearer`:

```diff
namespace Microsoft.AspNetCore.Authentication.JwtBearer;

public class JwtBearerOptions: AuthenticationSchemeOptions
{
+   public IList<TokenHandler> TokenHandlers { get; }
+   public bool UseTokenHandlers { get; set; } = true;
}
```

Additions in `WsFederation`:

```diff
namespace Microsoft.AspNetCore.Authentication.WsFederation;

public class WsFederationOptions: RemoteAuthenticationOptions
{
+   public IList<TokenHandler> TokenHandlers { get; }
+   public bool UseTokenHandlers { get; set; } = true;
}
```

Additions in `OpenIdConnect`:
We introduce a new boolean, `UseTokenHandler`, in the OpenIdConnect options to enable developers to decide whether the ID token validation will be done with the new `TokenHandler` (more performant, resilient and async) or with the legacy `SecurityTokenValidator`.

[PR for reference](https://github.com/dotnet/aspnetcore/pull/49333)

```diff
namespace Microsoft.AspNetCore.Authentication.OpenIdConnect;

public class OpenIdConnectOptions: RemoteAuthenticationOptions
{
+   public TokenHandler TokenHandler { get; }
+   public bool UseTokenHandler { get; set; } = true;
+   public bool MapInboundClaimsTokenHandler {get; set;}
}
```

## Usage Examples
By default, ASP.NET Core in .NET 8 uses the new TokenHandler. If a developer wants to use the legacy validators, they can set `UseTokenHandlers = false`. 

```csharp
services.Configure<JwtBearerOptions>(JwtBearerDefault.AuthenticationScheme, options => { options.UseTokenHandlers = false; });
```

If a developer wants to have their own TokenHandler, they can add it to the list of TokenHandlers:

```csharp
services.Configure<JwtBearerOptions>(JwtBearerDefault.AuthenticationScheme, options => { options.TokenHandlers.Add( new MyTokenHandler()); });
```

## Alternative Designs

Alternative designs were discussed with @Tratcher, @eerhardt, @halter73 . We went with Option A_1, but the alternatives discussed were:

**Option A_1**:
Have the same assemblies as today with a dual dependency on System.IdentityModel.Tokens.Jwt and Microsoft.IdentityModel.Tokens.JsonWebToken, and offer both interfaces (Jwt for compatibility, whereas the processing is done with JsonWebToken). In practice, note that the Jwt Wilson assembly already depends on the JsonWebToken Wilson assembly, so there would not be any additional dependencies than today.
Additionally, to leverage the new generation of Wilson assemblies without breaking changes, both ISecurityTokenValidator and TokenHandler members would have to be in ASP.NET core’s surface area.
Validation would need to be done on the options such that when using a new generation of Jwt classes, the new generation of the TokenHandlers would also need be used. 

**Option A_2**:
Duplicate the current ASP.NET Core assemblies, and have a new generation (let’s name them Microsoft.AspNetCore.Authentication.JwtBearer2. Microsoft.AspNetCore.Authentication.OpenIdConnect2, Microsoft.AspNetCore.Authentication.WsFederation2, Microsoft.AspNetCore.Authentication2 for now, until we have a better name), and have these new generation depend only on JsonWebToken, and only expose these concepts. Letting the old generation leverage only Jwt. Additionally, these new assemblies would only rely on TokenHandler and drop support for ISecurityTokenValidator
In practice, we could, for this Option A2, use the same codebase, but add the files of the old projects as links in the new project, with conditional projects)

**Option A_3**:
Breaking changes in ASP.NET to rely only on the Microsoft.IdentityModel.Tokens.JsonWebToken assembly for Jwts and only TokenHandler for token handler abstractions.
In practice, these breaking changes should only affect users that leverage the extensibility features. We need to understand how large of a population this would affect.

Also gathering customer feedback in the [GitHub discussion](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/discussions/2092) in Microsoft.IdentityModel repo and the two above mentioned PRs.

## Risks
When setting `UseTokenHandlers` or `UseTokenHandler` to `true`, the SecurityToken passed in the context of the TokenValidated event needs to be downcast to JsonWebToken instead of JwtSecurityToken for users who were already doing this, which is not a common scenario, but for more advanced users. Mitigation for the risk is to have an implicit operator.

Initial feedback on 7.0.0-preview of Microsoft.IdentityModel from [@kevinchalet](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/discussions/2092#discussioncomment-6440820): "FYI, I tested the 7.0.0-preview packages with OpenIddict and haven't seen any particular regression. Good job"

### PR DESCRIPTION
**Pull request by keegan-caruso**
# Move aspnetcore to leverage JsonWebToken and JsonWebTokenHandler in .NET 8 preview 7

Updates the JwtBearer, WsFederation and OpenIdConnect authentication handlers to leverage more
performant and resilient token validation. Also enables developers to opt-out of this new improvement
if they choose too in .NET 8.

## Description

ASP.NET, and then ASP.NET Core use **Microsoft.IdentityModel** to validate tokens. A more performant (30%), more resilient and async processing has been developed for three years and is widely used internally in Microsoft. It also offers defense in depth. This PR brings this improvement to ASP.NET Core in .NET 8 preview 7.

Fixes #49469 which also contains the API review

## Customer Impact

Customers can expect a 30% performance in token validation, and associated COGS improvements. They also benefit from
resiliency by the usage of a last-known-good cache of the OpenIdConnect metadata.

By default ASP.NET Core uses the new improved token validation. There is no breaking change to the public API, but a behavioral
change in the TokenValidatedContext classes (for the three authentication handlers): the `SecurityToken` property (of abstract type
SecurityToken) used to be backed by the previous generation JwtSecurityToken, and is now backed by another derived class JsonWebToken. Developers can opt-out by setting a boolean. For details see https://aka.ms/aspnetcore8/security-token-changes

## Regression?

- [ ] Yes
- [x] No

## Risk

- [ ] High
- [ ] Medium
- [x] Low

The new classes have been extensively exercised internally in Microsoft for several years.
The PR doesn't bring regression, based on community feedback (see for instance [here](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/discussions/2092#discussioncomment-6440820))

## Verification

- [ ] Manual (required)
- [x] Automated

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [ ] N/A

----

## When servicing release/2.1

- [ ] Make necessary changes in eng/PatchConfig.props

### CODE REVIEW

File: src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs
Comment (eerhardt) [inline src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs:145]:
```suggestion
    /// Gets or sets the <see cref="MapInboundClaims"/> property on the default instance of <see cref="JwtSecurityTokenHandler"/> in SecurityTokenValidators, or <see cref="JsonWebTokenHandler"/>, in TokenHandlers which is used when determining
```

File: src/Security/Authentication/test/JwtBearerTests_Handler.cs
Comment (eerhardt) [inline src/Security/Authentication/test/JwtBearerTests_Handler.cs:26]: Is there any way to refactor these tests so they can be shared with the existing JwtBearerTests? For example having a base test class with the tests in them?

May be something to think about as a follow-up refactoring.

File: src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs
Comment (eerhardt) [inline src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs:185]: I wonder if we should rethink these doc comments as they might not make full sense now that we've inverted the boolean to be `UseSecurityTokenValidators`.

Maybe we should add why someone would want to use SecurityTokenValidators as well.

File: src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs
Comment (eerhardt) [inline src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs:116]: @Tratcher @halter73 @captainsafia - what do you think about defining a new `DiagnosticId` for this obsoletion? That way people who are using SecurityTokenValidators, and want to keep using it, don't need to suppress `CS0618` (the catch-all obsolete ID), and instead can just globally suppress the specific `ASP1234` ID. Then we could do the same in our tests and not need to `#pragma warning disable` everywhere.

File: src/Security/Authentication/WsFederation/src/WsFederationHandler.cs
Comment (eerhardt) [inline src/Security/Authentication/WsFederation/src/WsFederationHandler.cs:386]: Why does the WsFederationHandler still cache the configuration in a field, but the JwtBearer doesn't?

See also the conversation at https://github.com/dotnet/aspnetcore/pull/48966#discussion_r1269654386.

File: src/Security/Authentication/OpenIdConnect/src/Resources.resx
Comment (BrennanConroy) [inline src/Security/Authentication/OpenIdConnect/src/Resources.resx:139]:
```suggestion
    <value>The Validated Security Token must be of type JsonWebToken, but instead its type is '{0}'.</value>
```

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs
Comment (BrennanConroy) [inline src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs:272]:
```suggestion
    /// This will be used instead of <see cref="SecurityTokenValidator"/> if <see cref="UseSecurityTokenValidator"/> is <see langword="false"/>.
```

File: src/Security/Authentication/WsFederation/src/WsFederationHandler.cs
Comment (keegan-caruso) [reply src/Security/Authentication/WsFederation/src/WsFederationHandler.cs:386]: We can make the same change to this handler if you want. It made sense to remove the field in JwtBearer to clean up the code. It was just being read in one place.

Was trying to make less changes to WsFed where possible.

File: src/Security/Authentication/test/JwtBearerTests_Handler.cs
Comment (keegan-caruso) [reply src/Security/Authentication/test/JwtBearerTests_Handler.cs:26]: agreed it would be a good follow up

File: src/Security/Authentication/OpenIdConnect/src/LoggingExtensions.cs
Comment (BrennanConroy) [inline src/Security/Authentication/OpenIdConnect/src/LoggingExtensions.cs:171]:
```suggestion
    [LoggerMessage(57, LogLevel.Error, "The Validated Security Token must be of type JsonWebToken, but instead its type is: '{SecurityTokenType}'.", EventName = "InvalidSecurityTokenTypeFromHandler")]
```

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs
Comment (BrennanConroy) [inline src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:743]:
```suggestion
                    tokenEndpointUser = ValidateToken(tokenEndpointResponse.IdToken, properties, validationParameters, out tokenEndpointJwt);
```

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs
Comment (BrennanConroy) [inline src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:1345]: Should there be a log here?

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs
Comment (BrennanConroy) [inline src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:1358]: Change this log method to accept `Type?` instead of string so we don't allocate the string unless it actually logs.

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs
Comment (BrennanConroy) [inline src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:1341]: Do the async methods in identity not accept a cancellation token?

File: src/Security/Authentication/WsFederation/src/WsFederationHandler.cs
Comment (BrennanConroy) [inline src/Security/Authentication/WsFederation/src/WsFederationHandler.cs:405]:
```suggestion
```

**Review by BrennanConroy — COMMENTED** [review]
Mainly nits since I'm not super familiar with the area

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs
Comment (keegan-caruso) [reply src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:1345]: It will be logged in catch block

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs
Comment (keegan-caruso) [reply src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:1341]: No

File: src/Security/Authentication/WsFederation/src/WsFederationHandler.cs
Comment (eerhardt) [reply src/Security/Authentication/WsFederation/src/WsFederationHandler.cs:386]: Ok. Let’s make a follow up change in main for this.

File: src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs
Comment (eerhardt) [reply src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs:116]: Let’s do this for rc1. @keegan-caruso - can you log a follow up issue for it?

**Review by eerhardt — APPROVED** [review]
The JwtBearer changes look good to me. It would be good to get others approval (for OIDC) before merge.

File: src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs
Comment (keegan-caruso) [reply src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs:116]: @eerhardt https://github.com/dotnet/aspnetcore/issues/49558

**Review by halter73 — APPROVED** [review]

**Pull request merged**
